File: src/main/java/me/hhitt/disasters/disaster/impl/MirrorControls.java
package me.hhitt.disasters.disaster.impl;

import me.hhitt.disasters.Disasters;
import me.hhitt.disasters.arena.Arena;
import me.hhitt.disasters.disaster.Disaster;
import me.hhitt.disasters.util.Notify;
import org.bukkit.Bukkit;
import org.bukkit.event.EventHandler;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerMoveEvent;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

public class MirrorControls implements Disaster, Listener {

    private final List<Arena> arenas = new ArrayList<>();
    private final Map<UUID, Long> guard = new ConcurrentHashMap<>();

    @Override
    public void start(Arena arena) {
        arenas.add(arena);
        Bukkit.getPluginManager().registerEvents(this, Disasters.getInstance());
        Notify.disaster(arena, "mirror-controls");
    }

    @Override
    public void pulse(int time) {
    }

    @Override
    public void stop(Arena arena) {
        arenas.remove(arena);
        HandlerList.unregisterAll(this);
    }

    @EventHandler
    public void onMove(PlayerMoveEvent event) {
        Arena arena = null;
        for (Arena a : arenas) {
            if (a.isPlayerValid(event.getPlayer())) {
                arena = a;
                break;
            }
        }
        if (arena == null) return;
        if (event.getFrom().getBlock().equals(event.getTo().getBlock())) return;
        long now = System.currentTimeMillis();
        Long last = guard.get(event.getPlayer().getUniqueId());
        if (last != null && now - last < 50) return;
        double dx = event.getTo().getX() - event.getFrom().getX();
        double dz = event.getTo().getZ() - event.getFrom().getZ();
        if (dx == 0.0 && dz == 0.0) return;
        org.bukkit.Location inverted = event.getFrom().clone();
        inverted.setX(event.getFrom().getX() - dx);
        inverted.setZ(event.getFrom().getZ() - dz);
        inverted.setYaw(event.getTo().getYaw());
        inverted.setPitch(event.getTo().getPitch());
        inverted.setY(event.getTo().getY());
        event.setTo(inverted);
        guard.put(event.getPlayer().getUniqueId(), now);
    }
}

File: src/main/kotlin/me/hhitt/disasters/disaster/impl/MirrorControls.kt
package me.hhitt.disasters.disaster.impl

import me.hhitt.disasters.arena.Arena
import me.hhitt.disasters.disaster.Disaster
import me.hhitt.disasters.util.Notify
import org.bukkit.Bukkit
import org.bukkit.event.EventHandler
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerMoveEvent
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * MirrorControls reverses horizontal displacement server-side.
 * Bukkit/Paper exposes movement/input reads but no raw client key rewrite setter,
 * so this disaster applies an inverted target via PlayerMoveEvent to simulate reversed controls.
 */
class MirrorControls : Disaster, Listener {

    private val arenas = mutableListOf<Arena>()
    private val guard = ConcurrentHashMap<UUID, Long>()

    override fun start(arena: Arena) {
        arenas.add(arena)
        Bukkit.getPluginManager().registerEvents(this, me.hhitt.disasters.Disasters.getInstance())
        Notify.disaster(arena, "mirror-controls")
    }

    override fun pulse(time: Int) {
    }

    override fun stop(arena: Arena) {
        arenas.remove(arena)
        HandlerList.unregisterAll(this)
    }

    @EventHandler
    fun onMove(event: PlayerMoveEvent) {
        val arena = arenas.firstOrNull { it.isPlayerValid(event.player) } ?: return
        if (event.from.block == event.to.block) return
        val now = System.currentTimeMillis()
        val last = guard[event.player.uniqueId] ?: 0L
        if (now - last < 50) return
        val dx = event.to.x - event.from.x
        val dz = event.to.z - event.from.z
        if (dx == 0.0 && dz == 0.0) return
        val inverted = event.from.clone()
        inverted.x = event.from.x - dx
        inverted.z = event.from.z - dz
        inverted.yaw = event.to.yaw
        inverted.pitch = event.to.pitch
        inverted.y = event.to.y
        event.to = inverted
        guard[event.player.uniqueId] = now
    }
}
